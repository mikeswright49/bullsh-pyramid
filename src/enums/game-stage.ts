export enum GameStage {
    Initiation,
    Memorization,
    Flipping,
    Declaration,
    Bullshit,
    Memory,
    Complete,
    Reveal,
}
