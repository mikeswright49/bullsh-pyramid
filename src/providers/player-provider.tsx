import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
