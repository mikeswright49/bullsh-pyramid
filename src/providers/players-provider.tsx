import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
