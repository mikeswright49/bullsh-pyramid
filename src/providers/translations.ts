export const TRANSLATIONS = {
    en: {
        'game.board.current.row': 'Current Row: {{ activeRow }}',
        'game.board.current.card': 'Current card',
        'game.display.stage.memorization.text': 'You have 30s to remember your cards!',
        'game.display.stage.flipping.text': 'The host is flipping a card',
        'game.display.stage.declaration.text': 'Let the world know if you have it',
        'game.display.stage.assign.text': 'Players are beginning to lie left right and center',
        'game.display.stage.bullshit.text': 'People should know who you think was lieing',
        'game.display.stage.reveal.text': 'We all now know the truth',
        'game.display.stage.memory.text': 'Well which cards did you have?',
        'game.display.stage.complete.text': 'Well time for a new game',
    },
};
