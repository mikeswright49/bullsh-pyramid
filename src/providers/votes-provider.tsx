import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
