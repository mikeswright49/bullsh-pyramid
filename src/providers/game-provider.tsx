import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
