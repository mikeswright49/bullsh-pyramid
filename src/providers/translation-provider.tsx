import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
