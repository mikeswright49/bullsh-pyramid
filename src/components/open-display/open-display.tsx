import React, { useContext } from 'react';
import { GameContext } from 'src/providers/game-provider';
import { PlayerContext } from 'src/providers/player-provider';
import { PlayersContext } from 'src/providers/players-provider';
import { GameStage } from 'src/enums/game-stage';
import { Hand } from '../hand/hand';
import { Card } from 'types/card';
import { PlayerStore } from 'src/stores/player-store';
import { TranslationContext } from 'src/providers/translation-provider';
import { PlayerInitiation } from '../player-display/stages/player-initiation';
import { PlayerMemorization } from '../player-display/stages/player-memorization';

const FLIP_TIMEOUT = 3000;
export function OpenDisplay() {
    const gameState = useContext(GameContext);
    const player = useContext(PlayerContext);
    const players = useContext(PlayersContext);
    const { translate } = useContext(TranslationContext);

    const { gameStage } = gameState;

    if (!player || !gameState) {
        return null;
    }

    function flipPlayerCard(card: Card) {
        card.hidden = false;
        PlayerStore.updatePlayer(player);
        setTimeout(() => {
            card.hidden = true;
            PlayerStore.updatePlayer(player);
        }, FLIP_TIMEOUT);
    }

    function renderOpenDisplay() {
        switch (gameStage) {
            case GameStage.Initiation:
                return <PlayerInitiation />;
            case GameStage.Memorization:
                return <PlayerMemorization />;
            case GameStage.Flipping:
                return (
                    <>
                        <div className="stack-y-2">
                            <h3>{translate('global.your_hand')}</h3>
                            <Hand cards={player.hand} onSelected={flipPlayerCard} />
                        </div>
                        {players
                            .filter((otherPlayer) => otherPlayer.id !== player.id)
                            .map((otherPlayer) => (
                                <div className="stack-y-2" key={otherPlayer.id}>
                                    <h5>
                                        {translate('global.player')}: {otherPlayer.name}
                                    </h5>
                                    <Hand cards={otherPlayer.hand} />
                                </div>
                            ))}
                    </>
                );
            default:
                return null;
        }
    }

    return <div data-testid="open-display">{renderOpenDisplay()}</div>;
}
