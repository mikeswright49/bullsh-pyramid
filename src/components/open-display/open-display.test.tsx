import React from 'react';
import { GameStage } from 'src/enums/game-stage';
import { OpenDisplay } from './open-display';
import { MOCK_PLAYER } from 'src/__tests__/data/mock-player';
import { render, screen } from '@testing-library/react';
import { GameContext } from 'src/providers/game-provider';
import { PlayerContext } from 'src/providers/player-provider';
import { GameState } from 'types/game-state';
import { MOCK_GAME_STATE } from 'src/__tests__/data/mock-game-state';
import { PlayersContext } from 'src/providers/players-provider';
import { MOCK_PLAYERS } from 'src/__tests__/data/mock-players';

describe('Open display', () => {
    let gameContextMock: GameState;

    beforeEach(() => {
        gameContextMock = Object.assign({}, MOCK_GAME_STATE);
    });

    const testCases = [[GameStage.Initiation], [GameStage.Memorization], [GameStage.Flipping]];
    test.each(testCases)('It should be able to render for %s', (stage) => {
        gameContextMock.gameStage = stage;
        render(
            <GameContext.Provider value={gameContextMock}>
                <PlayersContext.Provider value={MOCK_PLAYERS}>
                    <PlayerContext.Provider value={MOCK_PLAYER}>
                        <OpenDisplay />
                    </PlayerContext.Provider>
                </PlayersContext.Provider>
            </GameContext.Provider>
        );
        expect(screen.getByTestId('open-display')).toMatchSnapshot();
    });
});
