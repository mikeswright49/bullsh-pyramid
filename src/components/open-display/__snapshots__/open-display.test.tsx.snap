// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open display It should be able to render for 0 1`] = `
<div
  data-testid="open-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      player.display.stage.initiation.title
    </h3>
  </div>
</div>
`;

exports[`Open display It should be able to render for 1 1`] = `
<div
  data-testid="open-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      player.display.stage.memorization.title
    </h3>
    <div
      class="stack-y-2"
    >
      <h3>
        global.your_hand
      </h3>
      <div
        class="hand"
      >
        <div
          class="stack-x-4"
        >
          <div
            style="cursor: auto;"
          >
            <div
              class="flex-center-center card cardHidden"
              data-testid="card-2-7"
              style="background-position: -553px -246px;"
            />
          </div>
        </div>
        <div
          class="stack-x-4"
        >
          <div
            style="cursor: auto;"
          >
            <div
              class="flex-center-center card cardHidden"
              data-testid="card-3-8"
              style="background-position: -632px -369px;"
            />
          </div>
        </div>
        <div
          class="stack-x-4"
        >
          <div
            style="cursor: auto;"
          >
            <div
              class="flex-center-center card cardHidden"
              data-testid="card-1-11"
              style="background-position: -869px -123px;"
            />
          </div>
        </div>
        <div
          class="stack-x-4"
        >
          <div
            style="cursor: auto;"
          >
            <div
              class="flex-center-center card cardHidden"
              data-testid="card-2-0"
              style="background-position: -0px -246px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Open display It should be able to render for 2 1`] = `
<div
  data-testid="open-display"
>
  <div
    class="stack-y-2"
  >
    <h3>
      global.your_hand
    </h3>
    <div
      class="hand"
    >
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: pointer;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-7"
            style="background-position: -553px -246px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: pointer;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-3-8"
            style="background-position: -632px -369px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: pointer;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-1-11"
            style="background-position: -869px -123px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: pointer;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-0"
            style="background-position: -0px -246px;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="stack-y-2"
  >
    <h5>
      global.player
      : 
      joe
    </h5>
    <div
      class="hand"
    >
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: auto;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-1-4"
            style="background-position: -316px -123px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: auto;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-0-6"
            style="background-position: -474px -0px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: auto;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-2-2"
            style="background-position: -158px -246px;"
          />
        </div>
      </div>
      <div
        class="stack-x-4"
      >
        <div
          style="cursor: auto;"
        >
          <div
            class="flex-center-center card cardHidden"
            data-testid="card-0-8"
            style="background-position: -632px -0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
